lib/vector-core/src/event/metric/tags.rs

```diff
@@ -513,6 +526,18 @@ impl MetricTags {
     pub fn retain(&mut self, mut f: impl FnMut(&str, &mut TagValueSet) -> bool) {
         self.0.retain(|key, tags| f(key.as_str(), tags));
     }
+
+    /// Reduces all the tag values to their single value, discarding any for which that value would
+    /// be null.
+    pub(super) fn reduce_to_single(&mut self) {
+        self.retain(|_, values| match std::mem::take(values).into_single() {
+            Some(tag) => {
+                *values = TagValueSet::Single(TagValue::Value(tag));
+                true
+            }
+            None => false,
+        });
```

Comment:
nitpick: This might be a little bit clearer in two steps, like a `map` (or `filter_map`?) of the values `into_single` and then filter out empties. The current version is pretty dense.

Reply:
Done in 7b73ef952 and added a proptest to cover this as well.